examples/test_messaging/test_messaging.c

```diff
@@ -0,0 +1,391 @@
+        printf("Flags:\n");
+        printf(" - `-p <print_delay>`: number of packets between each print, e.g. `-p 1` prints every packets.\n");
+}
+
+/*
+ * Parse the application arguments.
+ */
+static int
+parse_app_args(int argc, char *argv[], const char *progname, struct onvm_nf *nf) {
+        int c;
+        struct test_msg_data *msg_params;
+        static struct rte_mempool *nf_msg_pool;
+
+        nf_msg_pool = rte_mempool_lookup(_NF_MSG_POOL_NAME);
+        if (nf_msg_pool == NULL)
+                rte_exit(EXIT_FAILURE, "No NF Message mempool - bye\n");
+
+        msg_params = (struct test_msg_data *)rte_malloc(NULL, sizeof(struct test_msg_data), 0);
+        msg_params->tests_passed = 0;
+        msg_params->tests_failed = 0;
+        msg_params->test_phase = 1;
+        msg_params->address = nf->service_id;
+        msg_params->msg_pool = nf_msg_pool;
+        msg_params->msg_q = nf->msg_q;
+        msg_params->ring_count = 0;
+        msg_params->test_msg_count = 0;
+        nf->data = (void *)msg_params;
+
+        while ((c = getopt(argc, argv, "p:")) != -1) {
+                switch (c) {
+                        case 'p':
+                                print_delay = strtoul(optarg, NULL, 10);
+                                RTE_LOG(INFO, APP, "print_delay = %d\n", print_delay);
+                                break;
+                        case '?':
+                                usage(progname);
+                                if (optopt == 'p')
+                                        RTE_LOG(INFO, APP, "Option -%c requires an argument.\n", optopt);
+                                else if (isprint(optopt))
+                                        RTE_LOG(INFO, APP, "Unknown option `-%c'.\n", optopt);
+                                else
+                                        RTE_LOG(INFO, APP, "Unknown option character `\\x%x'.\n", optopt);
+                                return -1;
+                        default:
+                                usage(progname);
+                                return -1;
+                }
+        }
+        return optind;
+}
+
+/*
+ * Sets up the NF. Initiates Unit Test
+ */
+void
+nf_setup(__attribute__((unused)) struct onvm_nf_local_ctx *nf_local_ctx) {
+        struct rte_mempool *pktmbuf_pool;
+
+        pktmbuf_pool = rte_mempool_lookup(PKTMBUF_POOL_NAME);
+        if (pktmbuf_pool == NULL) {
+                onvm_nflib_stop(nf_local_ctx);
+                rte_exit(EXIT_FAILURE, "Cannot find mbuf pool!\n");
+        }
+
+        printf("\nTEST MESSAGING STARTED\n");
+        printf("---------------------------\n");
+
+        struct test_msg_data *msg_params;
+
+        msg_params = (struct test_msg_data *)nf_local_ctx->nf->data;            
+
+        int *msg_int;
+        msg_int = (int*)rte_malloc(NULL, sizeof(int), 0);
+        *msg_int = MAGIC_NUMBER;
+        onvm_nflib_send_msg_to_nf(msg_params->address, msg_int);
```

Comment:
Does this function return a success/error code? If so you should check it here. Maybe that will give a clue at it falls after the first time. My guess is this is failing so the message is never sent and the other code is never triggered.

Reply:
I added that, but same issue. The problem comes after setup I believe, possibly in message handler.